File: backend/app/models/manufacturing.py
"""Manufacturing models — BOM, Work Orders, Workstations, Quality Checks."""
from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Any

from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
    func,
)
from sqlalchemy.dialects.postgresql import JSON, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin, UUIDPrimaryKeyMixin


# ── Bill of Materials ────────────────────────────────────────────────────────
class BillOfMaterials(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Recipe / formula for producing a finished item."""

    __tablename__ = "mfg_bom"

    bom_number: Mapped[str] = mapped_column(String(30), unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    finished_item_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_items.id"), nullable=False
    )
    quantity_produced: Mapped[int] = mapped_column(Integer, default=1)
    version: Mapped[int] = mapped_column(Integer, default=1)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    is_default: Mapped[bool] = mapped_column(Boolean, default=False)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    owner_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )

    items = relationship("BOMItem", back_populates="bom", cascade="all, delete-orphan", foreign_keys="[BOMItem.bom_id]")
    finished_item = relationship("InventoryItem", foreign_keys=[finished_item_id])


# ── BOM Item (line) ──────────────────────────────────────────────────────────
class BOMItem(UUIDPrimaryKeyMixin, Base):
    """Single raw material / sub-assembly line on a BOM."""

    __tablename__ = "mfg_bom_items"

    bom_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("mfg_bom.id"), nullable=False
    )
    item_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_items.id"), nullable=False
    )
    child_bom_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("mfg_bom.id"), nullable=True
    )
    quantity_required: Mapped[Decimal] = mapped_column(Numeric(12, 4), nullable=False)
    unit_of_measure: Mapped[str] = mapped_column(String(50), default="unit")
    scrap_percentage: Mapped[Decimal] = mapped_column(Numeric(5, 2), default=0)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    notes: Mapped[str | None] = mapped_column(String(500), nullable=True)

    bom = relationship("BillOfMaterials", back_populates="items", foreign_keys=[bom_id])
    item = relationship("InventoryItem", foreign_keys=[item_id])
    child_bom = relationship("BillOfMaterials", foreign_keys=[child_bom_id])


# ── WorkStation ──────────────────────────────────────────────────────────────
class WorkStation(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """A machine or work centre on the shop floor."""

    __tablename__ = "mfg_workstations"

    name: Mapped[str] = mapped_column(String(200), nullable=False)
    code: Mapped[str] = mapped_column(String(50), unique=True, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    capacity_per_hour: Mapped[Decimal | None] = mapped_column(Numeric(10, 2), nullable=True)
    hourly_rate: Mapped[Decimal] = mapped_column(Numeric(10, 2), default=0)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    warehouse_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_warehouses.id"), nullable=True
    )

    warehouse = relationship("Warehouse", foreign_keys=[warehouse_id])


# ── Work Order ───────────────────────────────────────────────────────────────
class WorkOrder(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Production work order."""

    __tablename__ = "mfg_work_orders"

    wo_number: Mapped[str] = mapped_column(String(30), unique=True, nullable=False)
    bom_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("mfg_bom.id"), nullable=False
    )
    workstation_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("mfg_workstations.id"), nullable=True
    )
    finished_item_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_items.id"), nullable=False
    )
    planned_quantity: Mapped[int] = mapped_column(Integer, nullable=False)
    completed_quantity: Mapped[int] = mapped_column(Integer, default=0)
    rejected_quantity: Mapped[int] = mapped_column(Integer, default=0)
    status: Mapped[str] = mapped_column(
        String(20), default="draft"
    )  # draft, planned, in_progress, completed, cancelled
    priority: Mapped[str] = mapped_column(
        String(10), default="medium"
    )  # low, medium, high
    planned_start: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    planned_end: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    actual_start: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    actual_end: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    target_warehouse_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_warehouses.id"), nullable=False
    )
    source_warehouse_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_warehouses.id"), nullable=False
    )
    total_material_cost: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    total_labor_cost: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    assigned_to: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    owner_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )

    bom = relationship("BillOfMaterials", foreign_keys=[bom_id])
    workstation = relationship("WorkStation", foreign_keys=[workstation_id])
    finished_item = relationship("InventoryItem", foreign_keys=[finished_item_id])
    target_warehouse = relationship("Warehouse", foreign_keys=[target_warehouse_id])
    source_warehouse = relationship("Warehouse", foreign_keys=[source_warehouse_id])
    materials = relationship("MaterialConsumption", back_populates="work_order", cascade="all, delete-orphan")
    quality_checks = relationship("QualityCheck", back_populates="work_order", cascade="all, delete-orphan")


# ── Material Consumption ─────────────────────────────────────────────────────
class MaterialConsumption(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Tracks raw material consumed by a work order."""

    __tablename__ = "mfg_material_consumption"

    work_order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("mfg_work_orders.id"), nullable=False
    )
    item_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_items.id"), nullable=False
    )
    planned_quantity: Mapped[Decimal] = mapped_column(Numeric(12, 4), nullable=False)
    actual_quantity: Mapped[Decimal] = mapped_column(Numeric(12, 4), default=0)
    warehouse_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_warehouses.id"), nullable=False
    )
    stock_movement_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_stock_movements.id"), nullable=True
    )
    consumed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    notes: Mapped[str | None] = mapped_column(String(500), nullable=True)

    work_order = relationship("WorkOrder", back_populates="materials")
    item = relationship("InventoryItem", foreign_keys=[item_id])
    warehouse = relationship("Warehouse", foreign_keys=[warehouse_id])


# ── Quality Check ────────────────────────────────────────────────────────────
class QualityCheck(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Quality inspection record for a work order."""

    __tablename__ = "mfg_quality_checks"

    check_number: Mapped[str] = mapped_column(String(30), unique=True, nullable=False)
    work_order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("mfg_work_orders.id"), nullable=False
    )
    inspector_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    checked_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    quantity_inspected: Mapped[int] = mapped_column(Integer, nullable=False)
    quantity_passed: Mapped[int] = mapped_column(Integer, nullable=False)
    quantity_failed: Mapped[int] = mapped_column(Integer, nullable=False)
    status: Mapped[str] = mapped_column(
        String(20), default="pending"
    )  # pending, passed, failed, partial
    parameters: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    work_order = relationship("WorkOrder", back_populates="quality_checks")
    inspector = relationship("User", foreign_keys=[inspector_id])


# ── Routing Step ────────────────────────────────────────────────────────────
class RoutingStep(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """A single operation/step in a production routing for a BOM."""

    __tablename__ = "mfg_routing_steps"

    bom_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("mfg_bom.id"), nullable=False
    )
    workstation_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("mfg_workstations.id"), nullable=False
    )
    sequence: Mapped[int] = mapped_column(Integer, nullable=False)
    operation: Mapped[str] = mapped_column(String(300), nullable=False)
    duration_minutes: Mapped[int] = mapped_column(Integer, nullable=False)

    bom = relationship("BillOfMaterials", foreign_keys=[bom_id])
    workstation = relationship("WorkStation", foreign_keys=[workstation_id])


# ── Scrap Entry ─────────────────────────────────────────────────────────────
class ScrapEntry(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Records scrapped / wasted items from a work order."""

    __tablename__ = "mfg_scrap_entries"

    work_order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("mfg_work_orders.id"), nullable=False
    )
    item_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False
    )
    quantity: Mapped[int] = mapped_column(Integer, nullable=False)
    reason: Mapped[str] = mapped_column(Text, nullable=False)
    date: Mapped[date] = mapped_column(Date, nullable=False)

    work_order = relationship("WorkOrder", foreign_keys=[work_order_id])


# ── Maintenance Schedule ────────────────────────────────────────────────────
class MaintenanceSchedule(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Scheduled maintenance for a workstation."""

    __tablename__ = "mfg_maintenance_schedules"

    workstation_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("mfg_workstations.id"), nullable=False
    )
    description: Mapped[str] = mapped_column(String(500), nullable=False)
    frequency: Mapped[str] = mapped_column(
        String(20), nullable=False
    )  # daily, weekly, monthly
    next_date: Mapped[date] = mapped_column(Date, nullable=False)
    last_completed: Mapped[date | None] = mapped_column(Date, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)

    workstation = relationship("WorkStation", foreign_keys=[workstation_id])


# ── Quality Control ─────────────────────────────────────────────────────────
class QualityControl(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Quality control test record for a work order."""

    __tablename__ = "mfg_quality_control"

    work_order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("mfg_work_orders.id"), nullable=False
    )
    test_name: Mapped[str] = mapped_column(String(300), nullable=False)
    result: Mapped[str] = mapped_column(
        String(20), nullable=False
    )  # pass, fail
    inspector_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    work_order = relationship("WorkOrder", foreign_keys=[work_order_id])
    inspector = relationship("User", foreign_keys=[inspector_id])

File: frontend/src/api/manufacturing.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import apiClient from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export interface BOMItem {
  id: string
  bom_id: string
  item_id: string
  child_bom_id: string | null
  quantity_required: number
  unit_of_measure: string
  scrap_percentage: number
  sort_order: number
  notes: string | null
  item_name?: string
}

export interface BOM {
  id: string
  bom_number: string
  name: string
  finished_item_id: string
  quantity_produced: number
  version: number
  is_active: boolean
  is_default: boolean
  notes: string | null
  owner_id: string
  created_at: string
  updated_at: string
  finished_item_name?: string
}

export interface BOMDetail extends BOM {
  items: BOMItem[]
}

export interface WorkStation {
  id: string
  name: string
  code: string
  description: string | null
  capacity_per_hour: number | null
  hourly_rate: number
  is_active: boolean
  warehouse_id: string | null
  created_at: string
  updated_at: string
}

export interface WorkOrder {
  id: string
  wo_number: string
  bom_id: string
  workstation_id: string | null
  finished_item_id: string
  planned_quantity: number
  completed_quantity: number
  rejected_quantity: number
  status: string
  priority: string
  planned_start: string | null
  planned_end: string | null
  actual_start: string | null
  actual_end: string | null
  target_warehouse_id: string
  source_warehouse_id: string
  total_material_cost: number
  total_labor_cost: number
  notes: string | null
  assigned_to: string | null
  owner_id: string
  created_at: string
  updated_at: string
  finished_item_name?: string
  bom_name?: string
}

export interface MaterialConsumption {
  id: string
  work_order_id: string
  item_id: string
  planned_quantity: number
  actual_quantity: number
  warehouse_id: string
  stock_movement_id: string | null
  consumed_at: string | null
  notes: string | null
  item_name?: string
}

export interface MaterialAvailability {
  item_id: string
  item_name: string | null
  sku: string | null
  required: number
  available: number
  sufficient: boolean
  shortfall: number
}

export interface QualityCheck {
  id: string
  check_number: string
  work_order_id: string
  inspector_id: string
  checked_at: string
  quantity_inspected: number
  quantity_passed: number
  quantity_failed: number
  status: string
  parameters: Record<string, unknown> | null
  notes: string | null
  created_at: string
  updated_at: string
}

export interface ManufacturingStats {
  total_boms: number
  total_workstations: number
  wo_draft: number
  wo_planned: number
  wo_in_progress: number
  wo_completed: number
  wo_cancelled: number
  in_progress_material_cost: number
  defect_rate_percent: number
}

export interface BOMCost {
  bom_id: string
  bom_number: string
  unit_cost: number
  total_cost: number
}

// ─── Payload types ────────────────────────────────────────────────────────────

export interface BOMItemPayload {
  item_id: string
  child_bom_id?: string | null
  quantity_required: number
  unit_of_measure?: string
  scrap_percentage?: number
  sort_order?: number
  notes?: string
}

export interface CreateBOMPayload {
  name: string
  finished_item_id: string
  quantity_produced?: number
  version?: number
  is_default?: boolean
  notes?: string
  items: BOMItemPayload[]
}

export interface UpdateBOMPayload {
  id: string
  name?: string
  finished_item_id?: string
  quantity_produced?: number
  version?: number
  is_active?: boolean
  is_default?: boolean
  notes?: string
  items?: BOMItemPayload[]
}

export interface CreateWorkStationPayload {
  name: string
  code: string
  description?: string
  capacity_per_hour?: number
  hourly_rate?: number
  warehouse_id?: string
}

export interface UpdateWorkStationPayload extends Partial<CreateWorkStationPayload> {
  id: string
  is_active?: boolean
}

export interface CreateWorkOrderPayload {
  bom_id: string
  workstation_id?: string
  planned_quantity: number
  priority?: string
  planned_start?: string
  planned_end?: string
  target_warehouse_id: string
  source_warehouse_id: string
  notes?: string
  assigned_to?: string
}

export interface UpdateWorkOrderPayload extends Partial<Omit<CreateWorkOrderPayload, 'bom_id'>> {
  id: string
}

export interface CreateQualityCheckPayload {
  work_order_id: string
  quantity_inspected: number
  quantity_passed: number
  quantity_failed: number
  status?: string
  parameters?: Record<string, unknown>
  notes?: string
}

export interface ConsumePayload {
  item_id: string
  quantity: number
  notes?: string
}

// ─── BOM hooks ────────────────────────────────────────────────────────────────

export function useBOMs(params: { search?: string; is_active?: boolean; skip?: number; limit?: number } = {}) {
  return useQuery({
    queryKey: ['manufacturing', 'bom', params],
    queryFn: async () => {
      const { data } = await apiClient.get<{ total: number; boms: BOM[] }>('/manufacturing/bom', { params })
      return data
    },
  })
}

export function useBOMDetail(id: string) {
  return useQuery({
    queryKey: ['manufacturing', 'bom', id],
    queryFn: async () => {
      const { data } = await apiClient.get<BOMDetail>(`/manufacturing/bom/${id}`)
      return data
    },
    enabled: !!id,
  })
}

export function useBOMCost(id: string) {
  return useQuery({
    queryKey: ['manufacturing', 'bom', id, 'cost'],
    queryFn: async () => {
      const { data } = await apiClient.get<BOMCost>(`/manufacturing/bom/${id}/cost`)
      return data
    },
    enabled: !!id,
  })
}

export function useCreateBOM() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: CreateBOMPayload) => {
      const { data } = await apiClient.post<BOMDetail>('/manufacturing/bom', payload)
      return data
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['manufacturing', 'bom'] }),
  })
}

export function useUpdateBOM() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ id, ...payload }: UpdateBOMPayload) => {
      const { data } = await apiClient.put<BOMDetail>(`/manufacturing/bom/${id}`, payload)
      return data
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['manufacturing', 'bom'] }),
  })
}

export function useDeleteBOM() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      await apiClient.delete(`/manufacturing/bom/${id}`)
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['manufacturing', 'bom'] }),
  })
}

// ─── WorkStation hooks ────────────────────────────────────────────────────────

export function useWorkStations() {
  return useQuery({
    queryKey: ['manufacturing', 'workstations'],
    queryFn: async () => {
      const { data } = await apiClient.get<WorkStation[]>('/manufacturing/workstations')
      return data
    },
  })
}

export function useCreateWorkStation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: CreateWorkStationPayload) => {
      const { data } = await apiClient.post<WorkStation>('/manufacturing/workstations', payload)
      return data
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['manufacturing', 'workstations'] }),
  })
}

export function useUpdateWorkStation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ id, ...payload }: UpdateWorkStationPayload) => {
      const { data } = await apiClient.put<WorkStation>(`/manufacturing/workstations/${id}`, payload)
      return data
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['manufacturing', 'workstations'] }),
  })
}

// ─── Work Order hooks ─────────────────────────────────────────────────────────

export function useWorkOrders(params: { status?: string; priority?: string; search?: string; skip?: number; limit?: number } = {}) {
  return useQuery({
    queryKey: ['manufacturing', 'work-orders', params],
    queryFn: async () => {
      const { data } = await apiClient.get<{ total: number; work_orders: WorkOrder[] }>('/manufacturing/work-orders', { params })
      return data
    },
  })
}

export function useWorkOrderDetail(id: string) {
  return useQuery({
    queryKey: ['manufacturing', 'work-orders', id],
    queryFn: async () => {
      const { data } = await apiClient.get<WorkOrder>(`/manufacturing/work-orders/${id}`)
      return data
    },
    enabled: !!id,
  })
}

export function useCreateWorkOrder() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: CreateWorkOrderPayload) => {
      const { data } = await apiClient.post<WorkOrder>('/manufacturing/work-orders', payload)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['manufacturing', 'work-orders'] })
      qc.invalidateQueries({ queryKey: ['manufacturing', 'dashboard'] })
    },
  })
}

export function useUpdateWorkOrder() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ id, ...payload }: UpdateWorkOrderPayload) => {
      const { data } = await apiClient.put<WorkOrder>(`/manufacturing/work-orders/${id}`, payload)
      return data
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['manufacturing', 'work-orders'] }),
  })
}

export function useStartWorkOrder() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      const { data } = await apiClient.post<WorkOrder>(`/manufacturing/work-orders/${id}/start`)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['manufacturing', 'work-orders'] })
      qc.invalidateQueries({ queryKey: ['manufacturing', 'dashboard'] })
      qc.invalidateQueries({ queryKey: ['inventory'] })
    },
  })
}

export function useCompleteWorkOrder() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ id, completed_quantity, rejected_quantity = 0 }: { id: string; completed_quantity: number; rejected_quantity?: number }) => {
      const { data } = await apiClient.post<WorkOrder>(
        `/manufacturing/work-orders/${id}/complete`,
        null,
        { params: { completed_quantity, rejected_quantity } },
      )
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['manufacturing', 'work-orders'] })
      qc.invalidateQueries({ queryKey: ['manufacturing', 'dashboard'] })
      qc.invalidateQueries({ queryKey: ['inventory'] })
    },
  })
}

export function useCancelWorkOrder() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      const { data } = await apiClient.post<WorkOrder>(`/manufacturing/work-orders/${id}/cancel`)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['manufacturing', 'work-orders'] })
      qc.invalidateQueries({ queryKey: ['manufacturing', 'dashboard'] })
    },
  })
}

export function useMaterialAvailability(woId: string) {
  return useQuery({
    queryKey: ['manufacturing', 'work-orders', woId, 'availability'],
    queryFn: async () => {
      const { data } = await apiClient.get<MaterialAvailability[]>(`/manufacturing/work-orders/${woId}/material-availability`)
      return data
    },
    enabled: !!woId,
  })
}

// ─── Material Consumption hooks ───────────────────────────────────────────────

export function useConsumeMaterial() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ woId, ...payload }: ConsumePayload & { woId: string }) => {
      const { data } = await apiClient.post<MaterialConsumption>(`/manufacturing/work-orders/${woId}/consume`, payload)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['manufacturing'] })
      qc.invalidateQueries({ queryKey: ['inventory'] })
    },
  })
}

export function useWorkOrderConsumption(woId: string) {
  return useQuery({
    queryKey: ['manufacturing', 'work-orders', woId, 'consumption'],
    queryFn: async () => {
      const { data } = await apiClient.get<MaterialConsumption[]>(`/manufacturing/work-orders/${woId}/consumption`)
      return data
    },
    enabled: !!woId,
  })
}

// ─── Quality Check hooks ──────────────────────────────────────────────────────

export function useQualityChecks(params: { work_order_id?: string; status?: string; skip?: number; limit?: number } = {}) {
  return useQuery({
    queryKey: ['manufacturing', 'quality-checks', params],
    queryFn: async () => {
      const { data } = await apiClient.get<{ total: number; quality_checks: QualityCheck[] }>('/manufacturing/quality-checks', { params })
      return data
    },
  })
}

export function useCreateQualityCheck() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: CreateQualityCheckPayload) => {
      const { data } = await apiClient.post<QualityCheck>('/manufacturing/quality-checks', payload)
      return data
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['manufacturing', 'quality-checks'] }),
  })
}

// ─── Dashboard ────────────────────────────────────────────────────────────────

export function useRemoveBOMItem() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ bomId, itemId }: { bomId: string; itemId: string }) => {
      await apiClient.delete(`/manufacturing/bom/${bomId}/items/${itemId}`)
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['manufacturing', 'bom'] }),
  })
}

export function useDeleteQualityCheck() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      await apiClient.delete(`/manufacturing/quality-checks/${id}`)
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['manufacturing', 'quality-checks'] }),
  })
}

export function useDeleteWorkStation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      await apiClient.delete(`/manufacturing/workstations/${id}`)
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['manufacturing', 'workstations'] }),
  })
}

export function useManufacturingStats() {
  return useQuery({
    queryKey: ['manufacturing', 'dashboard', 'stats'],
    queryFn: async () => {
      const { data } = await apiClient.get<ManufacturingStats>('/manufacturing/dashboard/stats')
      return data
    },
  })
}
